src/mgba_printf/mgba.c

```diff
@@ -0,0 +1,181 @@
+#define REG_DEBUG_FLAGS  ((vu16*) (0x4FFF700))
+#define REG_DEBUG_STRING ((char*) (0x4FFF600))
+
+#define PRINTF_BUFFER_SIZE (4096)
+
+static u32 MgbaConvertPStringToAsciiN(const char *src, char *dst, u32 n);
+static void MgbaPrintfBounded(s32 level, const char* ptr, ...);
+
+/*
+ * Converts a "PokeString" to ASCII. This function does not have any sanity checks for buffer lengths.
+ * Be careful not to overflow!
+ * Failed ASSERTIONS for buffers != NULL will halt the program!
+ */
+static u32 MgbaConvertPStringToAsciiN(const char *src, char *dst, u32 n)
+{
+    s32 i;
+
+    MGBA_ASSERT(dst != NULL);
+    MGBA_ASSERT(src != NULL);
+
+    for (i = 0; (*src != EOS) && i < n; i++, src++)
+    {
+        char modifiedCode = '?';
+        if (*src >= CHAR_a && *src <= CHAR_z)
+            modifiedCode = *src-(CHAR_a-'a'); // lower-case characters
+        else if (*src >= CHAR_A && *src <= CHAR_Z)
+            modifiedCode = *src-(CHAR_A-'A'); // upper-case characters
+        else if (*src >= CHAR_0 && *src <= CHAR_9)
+            modifiedCode = *src-(CHAR_0-'0'); // numbers
+        else if (*src == CHAR_SPACE)
+            modifiedCode = ' '; // space
+        else if (*src == CHAR_EXCL_MARK)
+            modifiedCode = '!'; // exclamation point
+        else if (*src == CHAR_QUESTION_MARK)
+            modifiedCode = '?'; // question mark
+        else if (*src == CHAR_PERIOD)
+            modifiedCode = '.'; // period
+        else if (*src == CHAR_DBL_QUOTE_LEFT || *src == CHAR_DBL_QUOTE_RIGHT)
+            modifiedCode = '"'; // double quote
+        else if (*src == CHAR_SGL_QUOTE_LEFT || *src == CHAR_SGL_QUOTE_RIGHT)
+            modifiedCode = '\''; // single quote
+        else if (*src == CHAR_CURRENCY)
+            modifiedCode = '$'; // currency mark (pokemonies in game, dollar sign in logs)
+        else if (*src == CHAR_COMMA)
+            modifiedCode = ','; // comma
+        else if (*src == CHAR_MULT_SIGN)
+            modifiedCode = '#'; // pound, hashtag, octothorpe, whatever
+        else if (*src == CHAR_SLASH)
+            modifiedCode = '/'; // slash
+        else if (*src == CHAR_LESS_THAN)
+            modifiedCode = '<'; // less than sign
+        else if (*src == CHAR_GREATER_THAN)
+            modifiedCode = '>'; // greater than sign
+        else if (*src == CHAR_PERCENT)
+            modifiedCode = '%'; // percentage
+        else if (*src == CHAR_LEFT_PAREN)
+            modifiedCode = '('; // opening parentheses
+        else if (*src == CHAR_RIGHT_PAREN)
+            modifiedCode = ')'; // closing parentheses
+        dst[i] = modifiedCode;
+    }
+    dst[i] = 0;
+    return i;
+}
+
+static void MgbaPrintfBounded(s32 level, const char* ptr, ...)
+{
+    va_list args;
+
+    level &= 0x7;
+    va_start(args, ptr);
+    mini_vsnprintf(REG_DEBUG_STRING, MGBA_REG_DEBUG_MAX, ptr, args);
+    va_end(args);
+    *REG_DEBUG_FLAGS = level | 0x100;
+}
+
+void MgbaPrintf(s32 level, const char* ptr, ...)
```

Comment:
this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.